sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/Sampler.java

```diff
@@ -77,5 +79,18 @@ SamplingResult shouldSample(
      *     {@link Decision#RECORD_AND_SAMPLE}.
      */
     Attributes getAttributes();
+
+    /**
+     * Return an optionally-updated {@link TraceState}, based on the parent TraceState. This may
+     * return the same {@link TraceState} that was provided originally, or an updated one.
+     *
+     * @param parentTraceState The TraceState from the parent span. Might be an empty TraceState, if
+     *     there is no parent. This will be the same TraceState that was passed in via the {@link
+     *     SpanContext} parameter on the {@link #shouldSample(SpanContext, String, String, Kind,
+     *     ReadableAttributes, List)} call.
+     */
+    default TraceState getUpdatedTraceState(TraceState parentTraceState) {
```

Comment:
Thanks sorry for missing the allocation issue, got it. TBH this API still feels weird to me - having to call the sampler, and then the caller has to correctly call this method to get the trace state. 

I'm reading the discussion here

https://github.com/open-telemetry/opentelemetry-java/pull/1707#discussion_r496019521

and I'm personally not sure this functional style is that much better than `null` indicating a non-update, in a caller

```
span.traceState = result.getUpdatedTraceState() != null ? result.getUpdatedTraceState() : parentTraceState;
```

seems more direct? It's true that there's no updated trace state, so use the non-updated one.

Otherwise, another approach that is similar to this one but seems a bit nicer is having this sort of method on the `Sampler`, accepting the decision with the parent trace state. "update trace state" is a business logic operation, it's nice if that lives in the Sampler which has the logic, a result feels like it should be mostly constant.

Finally, maybe we can just live with an extra allocation if it seems like a normal `getTraceState()` seems like a more intuitive API. Sampling result is very short lived, it may even be optimized away who knows.

Food for thought, I don't really have a strong preference for any, but a small preference for eating the allocation.

Reply:
The other reason to avoid "eating the allocation" is that we really want to make it super easy for "normal" samplers to do the right thing (that is, not modify the TraceState). I really dislike using "null" as a meaningful business result, but if y'all prefer that as the API, then I can get on board. What if we had it return `Optional<TraceState>` so the "normal" samplers can just have a constant `Optional.empty()`?

Reply:
@anuraaga and I discussed offline and decided to stick with what we have right now; if someone actually uses this API and doesn't like how it works, we can definitely change it, since it's SDK-only functionality, and we can be a little more loose with making changes.